census/vanilla-tf/trainer/task.py

```diff
@@ -0,0 +1,334 @@
+      for col in CATEGORICAL_COLS
+  ]
+
+  # convert the labels into one hot encoding
+  label_tensor = tf.one_hot(
+      label_df.astype('category').cat.codes.values,
+      2, off_value=1, on_value=0)
+
+  return (
+      continuous_columns + categorical_columns, label_tensor
+  )
+
+
+def sparse_cross(feature_tensors, num_buckets, name='cross'):
+  """Sparse feature cross of the feature SparseTensors."""
+  return sparse_feature_cross_op.sparse_feature_cross(
+      feature_tensors,
+      hashed_output=True,
+      num_buckets=num_buckets,
+      hash_key=tf.contrib.layers.SPARSE_FEATURE_CROSS_DEFAULT_HASH_KEY,
+      name=name
+  )
+
+
+def generate_wide_columns(input_columns):
+  """Generate wide columns by adding feature crosses of SparseTensors."""
+  (age, education_num, capital_gain, capital_loss, hours_per_week,
+   gender, race, education, marital_status, relationship, workclass,
+   occupation, native_country) = input_columns
+
+  # bucketize the age feature
+  # See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/ops/bucketization_op.py
+  age_bucket = bucketization_op.bucketize(age, [18, 25, 30, 35, 40, 45, 50, 55, 60, 65])
+
+  wide_columns = [
+      sparse_cross([education, occupation], 15 * 16, name='edu_occ'),
+      sparse_cross(
+        [age_bucket, race, occupation],
+        10 * 5 * 15,
+        name='age_race_occ'),
+      sparse_cross(
+        [native_country, occupation],
+        42 * 15,
+        name='native_country_occ'),
+      gender,
+      native_country,
+      education,
+      occupation,
+      workclass,
+      marital_status,
+      relationship,
+      age_bucket
+  ]
+
+  return wide_columns
+
+
+def concat_wide_columns(wide_columns):
+  """Concat the tensors from wide columns."""
+
+  (edu_occ, age_race_occ, native_country_occ, gender,
+   native_country, education, occupation, workclass,
+   marital_status, relationship, age_bucket) = wide_columns
+
+  dense_tensors = [
+      sparse_to_dense(edu_occ, 15 * 16),
+      sparse_to_dense(gender, 2),
+      sparse_to_dense(workclass, 9),
+      sparse_to_dense(native_country, 42),
+      sparse_to_dense(education, 16),
+      sparse_to_dense(occupation, 15),
+      sparse_to_dense(workclass, 9),
+      sparse_to_dense(marital_status, 7),
+      sparse_to_dense(relationship, 6)
+  ]
+
+  return tf.concat(dense_tensors, 1)
+
+
+def sparse_to_dense(sparse_tensor, vocab_size):
+  """Convert the sparse to dense tensor."""
+  dense_tensor = tf.sparse_to_indicator(sparse_tensor, vocab_size)
+  dense_tensor = tf.cast(dense_tensor, tf.int32)
+  return dense_tensor
+
+def read_input_tensor(input_file, skiprows=None):
+  """Concatenate the wide columns to produce a single tensor."""
+  inp, label = read_input_data(input_file, skiprows)
+  in_tensor, label_tensor = generate_input(inp, label)
+  return concat_wide_columns(generate_wide_columns(in_tensor)), label_tensor
+
+def loss(model, labels):
```

Comment:
I don't think `loss` and `optimizer` need methods. Just obsfucates the location of the hyperparameters.